app/controllers/v0/facilities/va_controller.rb

```diff
@@ -31,8 +53,61 @@ def suggested
            each_serializer: VASuggestedFacilitySerializer
   end
 
+  def provider_locator
+    ppms = Facilities::PPMSClient.new
+    providers = ppms.provider_locator(params)
+    Rails.logger.info(providers.class.name)
+    page = 1
+    page = Integer(params[:page]) if params[:page]
+    total = providers.length
+    start_ind = (page - 1) * BaseFacility.per_page
+    providers = providers[start_ind, BaseFacility.per_page - 1]
+    providers.map! do |provider|
+      prov_info = ppms.provider_info(provider['ProviderIdentifier'])
+      format_provloc(provider, prov_info)
+    end
+    # paging currently not possible
+    pages = { current_page: page, per_page: BaseFacility.per_page,
+              total_pages: total / BaseFacility.per_page, total_entries: total }
+
+    render json: { data: providers, meta: { pagination: pages } }
+  end
+
   private
 
+  def format_records(record, ppms)
+    if record.is_a?(BaseFacility)
+      ser = VAFacilitySerializer.new(record)
+      { id: ser.id, name: record[:name], attributes: ser.as_json }
+    else
+      prov_info = ppms.provider_info(record['ProviderIdentifier'])
+      format_provloc(record, prov_info)
+    end
+  end
+
+  def merge(facilities, providers, center_x, center_y, limit)
```

Comment:
as with my earlier comment... this feels like it's class/model level work .. can we think of a way to get it out of the controller and clean things up?

Reply:
Moved into the provider model.